src/Language/Spectacle/Syntax/Modal/Preterm.hs

```diff
@@ -0,0 +1,316 @@
+-- isomorphism:
+--
+-- @
+--                                                      'materialize'
+--                                   ───────────────────────────────────────────────>
+-- Members '[Modal, Logic] effs =>                                                      Preterm Bool
+--   Lang ctx effs Bool              <───────────────────────────────────────────────
+--                                                       'abstract'
+-- @
+--
+-- 'Preterm's are rewritten in the normalization process rather than the effects directly since interposing effect
+-- translations directly would be extremely complicated for the set of tautologies used to normalize temporal formula.
+--
+-- @since 0.1.0.0
+data Preterm a where
+  PreConst :: a -> Preterm a
+  PreConjunct :: Preterm a -> Preterm a -> Preterm a
+  PreDisjunct :: Preterm a -> Preterm a -> Preterm a
+  PreImplies :: Preterm a -> Preterm a -> Preterm a
+  PreNotImplies :: Preterm a -> Preterm a -> Preterm a
+  PreComplement :: Preterm a -> Preterm a
+  PreAlways :: Preterm a -> Preterm a
+  PreUpUntil :: Preterm a -> Preterm a -> Preterm a
+  deriving (Functor, Eq, Show)
+
+isComplement :: Preterm a -> Bool
+isComplement PreComplement {} = True
+isComplement _ = False
+{-# INLINE CONLIKE isComplement #-}
+
+-- | 'materialize' takes a 'Lang' with 'Modal' and 'Logic' effects and converts their representation to 'Preterm's.
+--
+-- @since 0.1.0.0
+materialize :: Lang ctx (Modal ': Logic ': effs) a -> Lang ctx effs (Preterm a)
+materialize = \case
+  Pure x -> pure (PreConst x)
+  Op op k -> case decomposeOp op of
+    Left op' -> case decomposeOp op' of
+      Left other -> Op other (materialize . k)
+      Right (Logic b) -> absurd b
+    Right (Modal b) -> absurd b
+  Scoped scoped loom -> case decomposeS scoped of
+    Left scoped' -> case decomposeS scoped' of
+      Left other -> Scoped other loom'
+      Right eff
+        | Conjunct lhs rhs <- eff -> do
+          lhs' <- runLoom loom' lhs
+          rhs' <- runLoom loom' rhs
+          return (PreConjunct lhs' rhs')
+        | Disjunct lhs rhs <- eff -> do
+           runLoom loom' lhs >>= \case
+            PreComplement lhs' -> do
+              rhs' <- runLoom loom' rhs
+              if isComplement rhs'
+                then return (PreDisjunct lhs' rhs')
+                else return (PreImplies lhs' rhs')
+            lhs' -> do
+              rhs' <- runLoom loom' rhs
+              return (PreDisjunct lhs' rhs')
+        | Complement expr <- eff -> do
+          expr' <- runLoom loom' expr
+          return (PreComplement expr')
+    Right eff
+      | Always expr <- eff -> do
+        expr' <- runLoom loom' expr
+        return (PreAlways expr')
+      | UpUntil lhs rhs <- eff -> do
+        lhs' <- runLoom loom' lhs
+        rhs' <- runLoom loom' rhs
+        return (PreUpUntil lhs' rhs')
+    where
+      loom' = loom ~>~ hoist materialize
```

Comment:
Nit-pick:

```suggestion
      reify = loom ~>~ hoist materialize
```

... what about a name like `reify` for the loom that hoists materialize?